### PR TITLE
PSC-5 Implemented Api router

### PR DESCRIPTION
**Pull request by MarkiyanVytrykush**

### CODE REVIEW

File: Uni/ApiRouter/DemoRouter.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import Foundation
+import UIKit
+
+// MARK: - HttpMethod
+enum HTTPMethod: String {
+    case get = "GET"
+}
+
+// MARK: - struct DemoRouter
+struct DemoRouter: ApiRoutable, DemoServiceProtocol {
```
Comment (RomanSlyepko) [inline Uni/ApiRouter/DemoRouter.swift]: Why `Demo`-router?

File: Uni/ApiRouter/ApiError.swift [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+//
+//  Created by Nanter on 8/27/19.
+//
+
+import Foundation
+import UIKit
+
+// MARK: -  ApiRouter
+enum ApiError: Error {
+    case unknow
```
Comment (RomanSlyepko) [inline Uni/ApiRouter/ApiError.swift]: typo

File: Uni/ApiRouter/ApiError.swift [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    case notRightURL
```
Comment (RomanSlyepko) [inline Uni/ApiRouter/ApiError.swift]: `invalidUrl` is a better name

File: Uni/ApiRouter/DemoRouter.swift (hunk 1 above)
Comment (RomanSlyepko) [inline Uni/ApiRouter/DemoRouter.swift]: conformance to `DemoServiceProtocol` is defined here, but it's not implemented in the current scope. Move conformance to the extension

File: Uni/ApiRouter/DemoRouter.swift [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    func asUrlRequest() throws -> URLRequest {
+        var urlRequ = URLComponents(string: baseURL)
```
Comment (RomanSlyepko) [inline Uni/ApiRouter/DemoRouter.swift]: 1. bad variable naming
2. nil `urlRequ` is not handled

File: Uni/ApiRouter/DemoRouter.swift [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    let headers: [String: String]
+    let body: Data?
+
+    // MARK: - init parametres
+    init(baseURL: String,
+         path: String,
+         method: HTTPMethod,
+         queryParameters: [String: String],
+         headers: [String: String],
+         body: Data?
```
Comment (RomanSlyepko) [inline Uni/ApiRouter/DemoRouter.swift]: default to `nil`

File: Uni/ApiRouter/PrototcolDemoRouter.swift [hunk 6]
```diff
@@ -0,0 +1,28 @@
+//
+//  PrototcolApiRouter.swift
```
Comment (RomanSlyepko) [inline Uni/ApiRouter/PrototcolDemoRouter.swift]: filename doesn't reflect the file contents

File: Uni/ApiRouter/ApiError.swift [hunk 7]
```diff
@@ -0,0 +1,18 @@
+//
+//  ApiError.swift
+//  Uni
+//
+//  Created by Nanter on 8/27/19.
```
Comment (RomanSlyepko) [inline Uni/ApiRouter/ApiError.swift]: must be a valid full name

File: Uni/ApiRouter/DemoRouter.swift (hunk 1 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/DemoRouter.swift]: fixed

File: Uni/ApiRouter/ApiError.swift (hunk 2 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/ApiError.swift]: fixed

File: Uni/ApiRouter/DemoRouter.swift (hunk 1 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/DemoRouter.swift]: fixed

File: Uni/ApiRouter/DemoRouter.swift (hunk 5 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/DemoRouter.swift]: fixed

File: Uni/ApiRouter/ApiError.swift (hunk 7 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/ApiError.swift]: fixed

File: Uni/ApiRouter/ApiError.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/ApiError.swift]: fixed

File: Uni/ApiRouter/ApiError.swift [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import Foundation
+import UIKit
+
+// MARK: -  ApiRouter
+enum ApiError: Error {
+    case unknow
+    case notRightURL
+}
+
+
```
[the remark this answers is not in the document]
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/ApiError.swift]: fixed

File: Uni/ApiRouter/PrototcolDemoRouter.swift (hunk 6 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/PrototcolDemoRouter.swift]: fixed

File: Uni/ApiRouter/DemoRouter.swift (hunk 4 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/DemoRouter.swift]: fixed

File: Uni/ApiRouter/ApiError.swift (hunk 3 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/ApiError.swift]: fixed

File: Uni/ApiRouter/DemoRouter.swift [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+// MARK: - struct DemoRouter
+struct DemoRouter: ApiRoutable {
+    func asUrlRequest() throws -> URLRequest {
+        guard var urlRequest = URLComponents(string: baseURL ) else {throw ApiError.unknown}
+        urlRequest.queryItems = queryParameters.map { URLQueryItem(name: $0.key, value: $0.value) }
+        guard let url = urlRequest.url else { throw ApiError.invalidUrl }
+        var request = URLRequest(url: url)
+        request.httpMethod = method.rawValue
+        for (key, value) in headers {
+            request.setValue(value, forHTTPHeaderField: key)
```
Comment (Alex-Chaku) [inline Uni/ApiRouter/DemoRouter.swift]: We spoke about code style yesterday, why it isn't fixed yet?
1. one line "for" with "forEach {" and usage of "$0"
2. "{throw ApiError.unknown}" needed spacing between brackets and code
3. URLComponents(string: baseURL ) exessive spaces after parameters

File: Uni/ApiRouter/PrototcolDemoRouter.swift [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+// MARK: - protocol ApiRoutable
+protocol ApiRoutable: UrlRequestConvertable {
+
+    var baseURL: String { get }
+    var path: String { get }
+    var method: HTTPMethod { get }
+    var queryParameters: [String: String] { get }
+    var headers: [String: String] { get }
+    var body: Data? { get }
```
Comment (Alex-Chaku) [inline Uni/ApiRouter/PrototcolDemoRouter.swift]: Where is default parameters and init?

File: Uni/ApiRouter/PrototcolDemoRouter.swift [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+    init(baseURL: String,
+         path: String,
+         method: HTTPMethod,
+         queryParameters: [String: String],
+         headers: [String: String],
+         body: Data? = nil) {
+        self.init(baseURL: baseURL,
+                 path: path,
+                 method: method,
```
Comment (Alex-Chaku) [inline Uni/ApiRouter/PrototcolDemoRouter.swift]: Where are default parameters for  
1. queryParameters: [String: String] = [:]
2. headers [String: String] = [:]

File: Uni/ApiRouter/ExtensionDemoRouter.swift [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+// MARK: - extension DemoRouter
+extension DemoRouter {
+    static func search(search: String, page: String) -> ApiRoutable {
+        return DemoRouter(baseURL: "https://api.songkick.com/",
+                          path: "v2/top-headlines",
+                          method: .get,
+                          queryParameters: [search: page],
+                          headers: [:],
+                          body: nil)
```
Comment (Alex-Chaku) [inline Uni/ApiRouter/ExtensionDemoRouter.swift]: If you have default parameters, why do you passing manually body: nil, headers: [:]

File: Uni/ApiRouter/PrototcolDemoRouter.swift (hunk 10 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/PrototcolDemoRouter.swift]: Fixed

File: Uni/ApiRouter/PrototcolDemoRouter.swift (hunk 11 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/PrototcolDemoRouter.swift]: Fixed

File: Uni/ApiRouter/ExtensionDemoRouter.swift (hunk 12 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/ExtensionDemoRouter.swift]: Fixed

File: Uni/ApiRouter/DemoRouter.swift (hunk 9 above)
Comment (MarkiyanVytrykush) [reply Uni/ApiRouter/DemoRouter.swift]: fixed

**Pull request closed without merging**